### ISSUE TITLE
Install over private network

### ISSUE THREAD

**Issue by minjatJ** [rancher/os#2779]
**RancherOS Version: (ros os version)**
Currently using 1.5.1 but any version I guess

**Where are you running RancherOS? (docker-machine, AWS, GCE, baremetal, etc.)**
Baremetal

Hi there,

Thanks for the amazing work, I have been using rancher since 1.2 and it keeps being great and better!

I am currently setting up a large volume installation process over a private network.
I am using iPXE +http server so I can easily provide any iso/vmlinuz/initrd/cloud-config etc...

Booting process is all fine and working well. But I want to install ros on those computers' disk.
For now the installation process is downloading rancher/os from github.com repository but I would like to avoid that and stay in my private network as much as possible.
I understand that the installation is looking from an iso on `/dev/sr0`
I am able to provide this iso (Custom rancheros.iso with preloaded images) over the network, I can mount it on `/dev/sr0` but it doesn't seem to be enough (I suppose it's because /dev/sr0 is already mounted)

Would you have any suggestion here ? I can PXE from the initrd of my custom Rancher OS on probably preload rancher/os in there but I am not sure it will work.

I have another question which might have been covered in https://github.com/rancher/os/issues/2697 (But I am not 100% clear on it). Is there anyway to add preloaded images during installation process to the system being installed (So I can access those images straightaway at first boot.)

Thanks for your time & help.
Joris

**Comment by niusmallnan** [comment]
If you want to autoformat the disk when booting by iPXE, you should add the autoformat part to kernel cmdline:
```
rancher.state.autoformat=[/dev/sda,/dev/vda] rancher.state.dev=LABEL=RANCHER_STATE rancher.state.wait
```
However, this does not install the bootloader to disk, so you cannot upgrade ROS.

---------

If you want to install completely to disk, you should not use `rancher.state.autoformat`.
You can push `rancher/os` image to your registry, then use `ros  install -i  myrepo/os:xxx` to install.

**Comment by minjatJ** [comment]
Hi @niusmallnan thanks for your answer

I indeed want to fully install ROS to the disk with the bootloader (The machine will be moved afterward with a less reliable network). I'll look into hosting ros image in my registry.

Also is it possible to copy preloaded images to a "soon to be installed" ros while installing it? (I would like the preloaded images to be on my system before the first boot, I guess I can just mount /dev/sdaX and put a tar of docker images in `/var/lib/rancher/preload/system-docker` but I'm wondering if there is another way)
Thanks!

**Comment by niusmallnan** [comment]
@minjatJ 
You can add a cloud-init datasource like `url`, then add this source to RancherOS via iPXE.

You can add some scripts via cloud-init to help install to disk automatically:
```
#cloud-config
write_files:
  - path: /etc/rc.local
    permissions: "0755"
    owner: root
    content: |
      #!/bin/bash
      wait-for-docker
      if [ -e /dev/disk/by-label/RANCHER_STATE ]; then
          # installed to disk
          ...
      else
          # not install to disk
          # install, but don't reboot
          # mount the STATE device, and save the preload images.
      fi
```

**Comment by minjatJ** [comment]
Yes I'am already doing it with the following config
```YAML
#cloud-config
hostname: rancherospxe
ssh_authorized_keys:
  - ssh-rsa AAAB....

write_files:
  - path: /etc/rc.local
    permissions: "0755"
    owner: root
    content: |
      #!/bin/bash
      (
      # ROS install will complain if /dev/sr0 doesn't exists
      mkdir /dev/sr0
      wait-for-docker
      echo y | ros install -f -c /cloud-config.yml -d /dev/sda
      ) &
  - path: /cloud-config.yml
    permissions: "0700"
    owner: root
    content: |
      #cloud-config
      hostname: freshros
      ssh_authorized_keys:
        - ssh-rsa AAAAB....
      write_files:
        - container: network
          path: /var/lib/iptables/eth1/rules.sh
          permissions: "0755"
          owner: root:root
          content: |
            #!/bin/bash
            set -ex
            # Open ports for rancher-agent registration
            iptables -I INPUT -p tcp --dport 6443 -j ACCEPT
            iptables -I INPUT -p tcp --dport 2376 -j ACCEPT
            # the last line of the file needs to be a blank line or a comment
        - path: /opt/install/install.sh
          permissions: "0544"
          owner: root
          content: |
            # Create VPN container
            system-docker run --restart=always --name vpn -d --cap-add=NET_ADMIN --net=host --device /dev/net/tun -v /opt/vpn:/vpn dperson/openvpn-client

            # Register rancher-agent
            docker run -d --privileged --restart=unless-stopped --net=host -v /etc/kubernetes:/etc/kubernetes -v /var/run:/var/run rancher/rancher-agent:v2.2.3 --server {.RancherServer} --token {.RancherAgentToken}  --etcd --controlplane --worker

            echo "VPN & Agent installed"
            touch /var/lib/rancher/install_on_disk.done
        - path: /etc/rc.local
          permissions: "0755"
          owner: root
          content: |
            #!/bin/bash
            (
            wait-for-docker

            # If vpn & agent are not installed yet
            if [ ! -e /var/lib/rancher/install_on_disk.done ];then
              /opt/install/install.sh >> /var/log/install.log 2>&1
            else
              # Need to manually start the system-docker vpn
              system-docker start vpn
              echo "All already installed" >>  /var/log/install.log
            fi
            ) &
        - path: /opt/vpn/vpn.ovpn
          owner: root
          permission: "0400"
          content: |
            client

            tls-client
            auth SHA256
            cipher AES-256-CBC
            remote-cert-tls server
            ....
      rancher:
        network:
          interfaces:
            l0:
            eth0:
              dhcp: true
            eth1:
              dhcp: true
              post_up:
              - /var/lib/iptables/eth1/rules.sh
```
I am not very happy with those `wait-for-docker` and manual checks but I can't find another way to do it. Would be great if we can wait for network (I tried `while ! wget etc...` but seems to be hanging)
Also after a manual reboot the system-docker container `vpn` (Started with `restart=always`) is not coming back automatically, I have to force a `system-docker start vpn`

Anyway those are minor issues that I should be able to fix soon. 
Thanks for your help!

**Comment by niusmallnan** [comment]
You can try `custom-system-services` for your vpn container. https://rancher.com/docs/os/v1.x/en/installation/system-services/custom-system-services/#labels

We don't have `wait-for-network`. Here are some you can refer to, I often use it.
```
    # check default routes
    for i in {1..10}; do
        # IPV4 UG: 0003
        # IPV6 UG: 00450003
        gateway=`cat /proc/net/route | grep 0003 | wc -l`
        gateway_v6=`cat /proc/net/ipv6_route | grep 00450003 | wc -l`
        if [[ $gateway -ge 1 || $gateway_v6 -ge 1 ]]; then
            ready="true"
            break
        else
            info "waiting for the default routes"
            sleep 1
        fi
    done
```

**Comment by minjatJ** [comment]
Yes I tried custom-system-services but I need access to the network in order to get the docker image but system services are launched before network (If I understand correctly). I also tried by create a service.yml in `/var/lib/rancher/conf/vpn.yml` then enable & up in rc.local but it seems that after a reboot I have to manually start it up again (Wondering if it's the intended behaviour or I might have miss something).

Thanks for the network snippet, that's very handy.
I'll close this ticket as I'm able to do pretty much everything I need.
Btw my company is trying to contact rancher but there isn't much reply so far (Slack/Forum/Contact page) any tip on who/where to contact for a potential high volume usage ?
